### PR TITLE
Create web socket backend for real-time editor (Collab Service)

### PR DESCRIPTION
**Pull request by weidak**
# Pull Request

## Description

Create a websocket backend to support real-time code collaboration

collab_service drawio (1)

## Related Issue(s)

#53 Create web socket for collaboration service

## Changes Made

Added logic into @tryyang2001's frontend implementation of the code editor, and collaboration service can be used across two browsers if you key in `http://<hosturl:port>/collaboration/room/<roomid>`  where so long as the room id is the same, users are able to collaborate.

## Screenshots (if applicable)

Not sure how to illustrate this LOL:
image

Connected status to websocket backend:
image

Disconnected status:
image

## Checklist
- [x] I have checked that the changes included in the PR are intended to merge to `master` or any destination branch.
- [x] I have verified that the new changes do not break any existing functionalities, unless the new changes are intended and have approved by the team.
- [x] I will take care of the merging and delete the side-branch after the PR is merged.

## Additional Notes/References
I have yet to add testing (not sure how to implement for this), and i have not done much when it comes to error handling because it requires implementation of caching. I think we should work on a caching/history service soon so we are able to restore code (either from backend/frontend) in the event of disconnection.

### CODE REVIEW

File: backend/collaboration-service/src/lib/enums/SocketEvent.ts
Comment (c0j0s) [inline backend/collaboration-service/src/lib/enums/SocketEvent.ts:1]: 👍

File: frontend/src/types/enums.ts [hunk 1]
```diff
@@ -81,3 +76,11 @@ export enum ToastType {
   WARNING = "warning",
   INFO = "info",
 }
+
+export enum SocketEvent {
+  CONNECT = "connect",
+  DISCONNECT = "disconnect",
+  JOIN_ROOM = "join_room",
+  CODE_CHANGE = "code_change",
+  CODE_UPDATE = "code_update",
```
Comment (c0j0s) [inline frontend/src/types/enums.ts:85]: Will it cause any issues if I include matching socket event here? if ok, then i'll add it in 😊

File: frontend/src/components/collab/CodeEditorPanel.tsx [hunk 2]
```diff
@@ -16,18 +18,48 @@ const CodeEditorPanel: FC<CodeEditorPanelProps> = ({
...
+    if (socketService) {
+      socketService.receiveCodeUpdate(setCurrentCode);
+    }
+  });
+
+  setInterval(() => {
+    if (socketService) {
+      setIsSocketConnected(socketService.getConnectionStatus());
+    }
+  })
```
Comment (c0j0s) [inline frontend/src/components/collab/CodeEditorPanel.tsx:42]: would this cause performance issue?

**Review by c0j0s — COMMENTED** [review]
Generally LGTM, just question with getMatchedRecord, why would the code editor fetch record from matching service? I thought the matching service should redirect the user to the code editor page with the required properties instead....

File: frontend/src/types/enums.ts (hunk 1 above)
Comment (weidak) [reply frontend/src/types/enums.ts:85]: No issues! Pls do HAHAHA

File: frontend/src/components/collab/CodeEditorPanel.tsx (hunk 2 above)
Comment (weidak) [reply frontend/src/components/collab/CodeEditorPanel.tsx:42]: I'm not certain actually... i dont notice anything but maybe ill set it to poll every 0.5/1s instead

**Comment by weidak** [comment]
@c0j0s ah i'm not too sure about how the params will be passed (so we are using a dummy wrapper rn). i think we would need quite a bit of information, so i wasnt sure if the router will be able to pass all these through the query? if can we will change it accordingly but i think we should discuss this during the meeting tomorrow

**Review by c0j0s — APPROVED** [review]
LGTM, solid👍

**Comment by weidak** [comment]
I think since leyi will be working on the chatting function in the same service, pls do take a look at my implementation too :D

**Comment by tryyang2001** [comment]
> Why would the code editor fetch record from matching service?

To initialise the code editor, we need several information. For example, I need to know which question is given so that I can initialise the code template. I also need to know who is your partner, which language did you all get. Furthermore, as I believe in our concurrent coding implementation, we create sockets which need a specific id (let's call it roomId), so we also need a roomId.

Overall, I am not sure how I can get all these information, how I foresee this is that when the user starts a match, and if the match is successful, we need to redirect the user to the workspace page. Whether the information needed is passed through functional params/properties, or through a GET API call to matching API are both fine. As @weidak mentioned that we can change accordingly. So feel free to use your current implementation in your matching service, we will try to follow your way as much as possible!

**Review by tlyi — APPROVED** [review]
LGTM! Thank you 👍 😺

**Pull request merged**